### PR TITLE
Upgrade to LSP 3.16

### LINKED ISSUE #110: Update jsonrpc and languageserver dependencies
New updates have come for vscode dependencies. We need to apply the upgrades. The new languageserver-protocol has new features that we should add to atom-languageclient.

https://microsoft.github.io/language-server-protocol/specifications/specification-current/#version_3_16_0
https://github.com/microsoft/vscode-languageserver-node#3160-protocol-600-json-rpc-700-client-and-700-server
```
 vscode-jsonrpc                   5.0.1  →   6.0.0
 vscode-languageserver-protocol  3.15.3  →  3.16.0
 vscode-languageserver-types     3.15.1  →  3.16.0
```

### PR DESCRIPTION
**Pull request by illright**
I'm currently working on upgrading the LSP version. For now I have just set up some foundations, fixed compiler errors from the update, and now I'm gonna start implementing the new features. Figured it'd be good to get feedback as I'm working since this is my first dive into the Atom ecosystem

Closes #110

### CODE REVIEW

File: lib/adapters/show-document-adapter.ts
Comment (illright) [inline lib/adapters/show-document-adapter.ts:2]: @aminya This is the line that I asked you in Discord about

File: lib/adapters/show-document-adapter.ts
Comment (illright) [inline lib/adapters/show-document-adapter.ts:35]: @aminya TypeScript trips up here because the DefinitelyTyped types don't mention that `atom.workspace.open` returns a `Promise<TextEditor>`. I've submitted a PR to that repo, you should be mentioned to review it. That will fix the other CI error

File: lib/adapters/show-document-adapter.ts
Comment (UziTech) [inline lib/adapters/show-document-adapter.ts:40]: Maybe instead of opening it here we could send it to the client and the client will handle opening the url?

Comment (illright) [reply lib/adapters/show-document-adapter.ts:40]: How can we send it? Define a callback on `AutoLanguageClient`? Moreover, that'd mean doing the same thing over and over in every language client, we should avoid this if possible

File: lib/adapters/show-document-adapter.ts
Comment (aminya) [reply lib/adapters/show-document-adapter.ts:2]: Adding electron.d.ts fixes this
```ts
// Electron types
declare module 'electron' {
  export const shell: {
    openExternal(url: string): void
  }
}


```

File: lib/adapters/show-document-adapter.ts
Comment (aminya) [reply lib/adapters/show-document-adapter.ts:40]: I am fine with opening it here.

Comment (UziTech) [reply lib/adapters/show-document-adapter.ts:40]: I'm not sure if I am reading the spec correctly but it seems according to the spec that the uri is supposed to be a document that should be opened in Atom not externally.

Comment (UziTech) [reply lib/adapters/show-document-adapter.ts:40]:
```suggestion
    shell.openExternal(params.uri, { activate: params.takeFocus })
```

File: lib/adapters/show-document-adapter.ts
Comment (aminya) [reply lib/adapters/show-document-adapter.ts:35]: Ongoing PR:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50382

File: lib/adapters/show-document-adapter.ts
Comment (illright) [reply lib/adapters/show-document-adapter.ts]: Yep, didn't notice it in because of the TypeScript error. Fixed now.

File: lib/adapters/show-document-adapter.ts
Comment (illright) [reply lib/adapters/show-document-adapter.ts:34]: Alright, worked around it.

**Comment by UziTech** [comment]
I merged @types/atom v1.40.6 if you rebase that should help with typescript.

**Comment by aminya** [comment]
Can we get the tests working and merge this for now? We can add other features in separate PRs.

**Comment by UziTech** [comment]
I think we should also get tests written for this before merging.

**Comment by illright** [comment]
Yeah, I'll polish it up soon

**Comment by aminya** [comment]
Any update on this?

**Comment by illright** [comment]
I'm a bit lost with the tests, honestly, because I could simply fix the failing tests and be done with it, but we wanna do it properly and create tests for new functionality. That's where I fall into the endless loophole of trying to create a thorough test suite, realizing that the surrounding tests aren't thorough enough and/or need to be updated. 

Besides, a lot of stuff isn't finished here, and I don't even know if this PR is of any value as of now, apart from all the scaffolding that is meant to become functionality in the future. Is there a reason why you want to merge this soon?

**Comment by aminya** [comment]
When a PR becomes big, it is hard to review and test, and it also blocks other improvements. 
We want to continue developing this package, and because this PR is editing multiple files, it is hard to do this without introducing conflicts.

**Comment by UziTech** [comment]
> tests for the new features in other pull requests

No we should not merge this without proper tests. The tests make it easier to review this since we can know that it is working the way it is supposed to.

**Comment by aminya** [comment]
It should only add tests for the new features. I suggest reverting the new features, so we can merge this.

**Comment by aminya** [comment]
I added a PR (#122) which bumps the LSP. 

Please cherry-pick the separate features, and open separate PRs for each (including the tests). This byte size approach makes everything easier and reduces the risk of bugs.

**Comment by UziTech** [comment]
I agree, smaller PRs are easier to test and review, but we still need tests for any new features and updates.

**Comment by aminya** [comment]
So far these are the changes I have been able to extract out of this PR. I have added the tests for the new features.

- #122 
- #135 
- #136 
- #137 
- #138

**Review by aminya — CHANGES_REQUESTED** [review]
Thanks for your contributions. 

I am closing this PR as I have been able to extract the following PRs out of this. Various modifications were applied, and I have added the tests for the new features.

- #122 
- #135 
- #136 
- #137 
- #138 
- #139

As a suggestion for your future contributions, if you can divide your contributions into separate small incremental changes, it would make it much easier for us to review, test, and verify. As an example, this PR had [a hidden change](https://github.com/atom-community/atom-languageclient/pull/136#discussion_r597515647) that could break the whole Autocomplete for the clients.

**Pull request closed without merging**
